<!DOCTYPE html><html xmlns="http://www.w3.org/1999/xhtml"><head><title>Error Handling</title><meta name="generator" content="Hovercraft! 1.0 http://regebro.github.com/hovercraft"></meta><link rel="stylesheet" href="css/hovercraft.css" media="all"></link><link rel="stylesheet" href="css/impressConsole.css" media="all"></link><link rel="stylesheet" href="css/highlight.css" media="all"></link><link rel="stylesheet" href="CSS/course.css" media="screen,projection"></link></head><body class="impress-not-supported"><div id="impress"><div class="step step-level-1" step="0" data-rotate-x="0" data-rotate-y="0" data-rotate-z="0" data-scale="1" data-x="0" data-y="0" data-z="0"><h1 id="error-handling">Error Handling</h1><p>&#x201C;Don't interrupt me while I'm interrupting."
&#x2015; Winston S. Churchill</p></div><div class="step step-level-1" step="1" data-rotate-x="0" data-rotate-y="0" data-rotate-z="0" data-scale="1" data-x="1600" data-y="0" data-z="0"><h1 id="outline">Outline</h1><p>Mostly we will be looking at Exception Handling.</p><ul><li>What is an exception<ul><li>try catch</li><li>throw</li><li>exception guarantees</li></ul></li><li>RAII and scopeguard</li><li>Traditional Error Handling</li></ul></div><div class="step step-level-1" step="2" data-rotate-x="0" data-rotate-y="0" data-rotate-z="0" data-scale="1" data-x="3200" data-y="0" data-z="0"><h1 id="introduction">Introduction</h1><p>Will we focus on the use of exceptions for errors that can't be handled locally.
I will refer to "a library" as any code that is invoked to preform a task.</p><ul><li>A library is just normal code, but often the writer doesn't know how their code will be used.</li><li>The Author of a library can often detect a run-time error but does not in general have any idea what to do about it.</li><li>The User of a library may know how to deal with an error but can't easily detect it.</li></ul><pre class="highlight code c++"><span class="kt">void</span> <span class="nf">taskmaster</span><span class="p">()</span> <span class="p">{</span>
    <span class="k">try</span><span class="p">{</span>
      <span class="k">auto</span> <span class="n">result</span> <span class="o">=</span> <span class="n">do_task</span><span class="p">();</span>
      <span class="c1">//use result;
</span>    <span class="p">}</span>
    <span class="k">catch</span><span class="p">(</span><span class="n">Some_Error</span><span class="p">)</span> <span class="p">{</span>
      <span class="c1">//Failure to do_task: Handle Problem.
</span>    <span class="p">}</span>
<span class="p">}</span>

<span class="kt">int</span> <span class="nf">do_task</span><span class="p">(){</span>
    <span class="c1">//..do something
</span>    <span class="k">if</span><span class="p">(</span><span class="cm">/*could preform task*/</span><span class="p">)</span>
      <span class="k">return</span> <span class="n">result</span><span class="p">;</span>
    <span class="k">throw</span> <span class="n">Some_Error</span><span class="p">{};</span>
<span class="p">}</span></pre></div><div class="step step-level-1" step="3" data-rotate-x="0" data-rotate-y="0" data-rotate-z="0" data-scale="1" data-x="4800" data-y="0" data-z="0"><h1 id="traditional-approaches">Traditional Approaches</h1><ul><li><strong>Terminate the Program</strong>: pretty drastic approach and often we can do better</li></ul><pre class="highlight code c++"><span class="k">if</span><span class="p">(</span><span class="n">something_wrong</span><span class="p">)</span> <span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">);</span></pre><ul><li><strong>Return an error value</strong>: Not always feasible. Very much C style.</li></ul><pre class="highlight code c++"><span class="kt">int</span> <span class="nf">getInt</span><span class="p">();</span> <span class="o">-&gt;</span> <span class="kt">int</span> <span class="n">getInt</span><span class="p">(</span><span class="kt">int</span><span class="o">&amp;</span> <span class="n">out</span><span class="p">);</span></pre><ul><li><strong>Return a legal value and leave the program in an ""error state.""</strong><ul><li>many C library functions set the nonlocal variable errno to indicate an error.</li><li>Easy to forget to check.</li></ul></li><li><strong>Call an error-handler functino</strong><ul><li>Unless error-handler can fully handle the error likely still should call terminate.</li></ul></li></ul><pre class="highlight code c++"><span class="k">if</span><span class="p">(</span><span class="n">something_wrong</span><span class="p">)</span> <span class="n">something_handler</span><span class="p">();</span></pre></div><div class="step step-level-1" step="4" data-rotate-x="0" data-rotate-y="0" data-rotate-z="0" data-scale="1" data-x="6400" data-y="0" data-z="0"><h1 id="exceptions-default-to-terminate">Exceptions Default to Terminate</h1><ul><li>Unhanded errors call terminate</li><li>Exception handling makes programs more "brittle"<ul><li>This sounds bad but has some advantages.</li><li>Preferable to getting wrong results later in the development process or after release.</li><li>We can catch all exceptions when termination is unacceptable.</li></ul></li></ul></div><div class="step step-level-1" step="5" data-rotate-x="0" data-rotate-y="0" data-rotate-z="0" data-scale="1" data-x="8000" data-y="0" data-z="0"><h1 id="hierarchical-error-handling">Hierarchical Error Handling</h1><ul><li>Successful fault-tolerant systems are multi-level<ul><li>Each level copes with as many errors as it can without getting too contorted.</li><li>Not every function should be a firewall.</li><li>Leaves exceptions for higher levels if the current level can't deal with them</li></ul></li><li>This will be better understood when doing todays lab.</li></ul></div><div class="step step-level-1" step="6" data-rotate-x="0" data-rotate-y="0" data-rotate-z="0" data-scale="1" data-x="9600" data-y="0" data-z="0"><img src="images/showCode.jpg"></img></div><div class="step step-level-1" step="7" data-rotate-x="0" data-rotate-y="0" data-rotate-z="0" data-scale="1" data-x="11200" data-y="0" data-z="0"><h1 id="catching-exceptions">Catching Exceptions</h1><pre class="highlight code c++"><span class="kt">void</span> <span class="nf">foo</span><span class="p">()</span> <span class="p">{</span>
    <span class="k">try</span><span class="p">{</span>
        <span class="c1">//do stuff
</span>    <span class="p">}</span>
    <span class="k">catch</span><span class="p">(</span><span class="k">const</span> <span class="n">std</span><span class="o">::</span><span class="n">runtime_error</span><span class="o">&amp;</span> <span class="n">e</span><span class="p">){</span>
       <span class="c1">//Handle Runtime Error
</span>    <span class="p">}</span>
    <span class="k">catch</span><span class="p">(</span><span class="k">const</span> <span class="n">std</span><span class="o">::</span><span class="n">exception</span><span class="o">&amp;</span> <span class="n">e</span><span class="p">){</span>
       <span class="c1">//Handle every standard-library exception
</span>    <span class="p">}</span>
    <span class="k">catch</span><span class="p">(...){</span>
       <span class="c1">//Handle anything.. but you don't know what it is. :(
</span>    <span class="p">}</span>
<span class="p">}</span></pre></div><div class="step step-level-1" step="8" data-rotate-x="0" data-rotate-y="0" data-rotate-z="0" data-scale="1" data-x="12800" data-y="0" data-z="0"><h1 id="id1">Catching Exceptions</h1><p>The more specific handlers get bypassed.</p><pre class="highlight code c++"><span class="kt">void</span> <span class="nf">foo</span><span class="p">()</span> <span class="p">{</span>
    <span class="k">try</span><span class="p">{</span>
        <span class="c1">//do stuff
</span>    <span class="p">}</span>
    <span class="k">catch</span><span class="p">(...){</span>
       <span class="c1">//Handle anything.. but you don't know what it is. :(
</span>    <span class="p">}</span>
    <span class="k">catch</span><span class="p">(</span><span class="k">const</span> <span class="n">std</span><span class="o">::</span><span class="n">exception</span><span class="o">&amp;</span> <span class="n">e</span><span class="p">){</span>
       <span class="c1">//Handle Generic std::exception
</span>    <span class="p">}</span>
    <span class="k">catch</span><span class="p">(</span><span class="k">const</span> <span class="n">std</span><span class="o">::</span><span class="n">runtime_error</span><span class="o">&amp;</span> <span class="n">e</span><span class="p">){</span>
       <span class="c1">//Handle Runtime Error
</span>    <span class="p">}</span>
<span class="p">}</span></pre></div><div class="step step-level-1" step="9" data-rotate-x="0" data-rotate-y="0" data-rotate-z="0" data-scale="1" data-x="14400" data-y="0" data-z="0"><h1 id="exceptions-in-std">Exceptions in std::</h1><img src="images/stdExceptions.png"></img></div><div class="step step-level-1" step="10" data-rotate-x="0" data-rotate-y="0" data-rotate-z="0" data-scale="1" data-x="16000" data-y="0" data-z="0"><h1 id="function-try-block">Function-try-block</h1><ul><li>Primarily used for logging.</li><li>Can't access members of the class or of a base.</li><li>Every catch-clause in the function must terminate by throwing.</li></ul><pre class="highlight code c++"><span class="k">struct</span> <span class="n">S</span> <span class="p">{</span>
    <span class="n">std</span><span class="o">::</span><span class="n">string</span> <span class="n">m</span><span class="p">;</span>
    <span class="n">S</span><span class="p">(</span><span class="k">const</span> <span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="o">&amp;</span> <span class="n">arg</span><span class="p">)</span> <span class="k">try</span> <span class="o">:</span> <span class="n">m</span><span class="p">(</span><span class="n">arg</span><span class="p">,</span> <span class="mi">100</span><span class="p">)</span> <span class="p">{</span>
        <span class="n">std</span><span class="o">::</span><span class="n">cout</span> <span class="o">&lt;&lt;</span> <span class="s">"constructed, mn = "</span> <span class="o">&lt;&lt;</span> <span class="n">m</span> <span class="o">&lt;&lt;</span> <span class="sc">'\n'</span><span class="p">;</span>
    <span class="p">}</span> <span class="k">catch</span><span class="p">(</span><span class="k">const</span> <span class="n">std</span><span class="o">::</span><span class="n">exception</span><span class="o">&amp;</span> <span class="n">e</span><span class="p">)</span> <span class="p">{</span>
        <span class="n">std</span><span class="o">::</span><span class="n">cerr</span> <span class="o">&lt;&lt;</span> <span class="s">"arg="</span> <span class="o">&lt;&lt;</span> <span class="n">arg</span> <span class="o">&lt;&lt;</span> <span class="s">" failed: "</span> <span class="o">&lt;&lt;</span> <span class="n">e</span><span class="p">.</span><span class="n">what</span><span class="p">()</span> <span class="o">&lt;&lt;</span> <span class="sc">'\n'</span><span class="p">;</span>
    <span class="p">}</span> <span class="c1">// implicit throw; here
</span><span class="p">};</span></pre></div><div class="step step-level-1" step="11" data-rotate-x="0" data-rotate-y="0" data-rotate-z="0" data-scale="1" data-x="17600" data-y="0" data-z="0"><h1 id="rethrowing-exceptions">Rethrowing Exceptions</h1><p>I can't fully clean up but I want to do something.</p><pre class="highlight code c++"><span class="kt">void</span> <span class="nf">foo</span><span class="p">()</span> <span class="p">{</span>
    <span class="k">try</span><span class="p">{</span>
        <span class="c1">//do stuff
</span>    <span class="p">}</span>
    <span class="k">catch</span><span class="p">(</span><span class="k">const</span> <span class="n">std</span><span class="o">::</span><span class="n">runtime_error</span><span class="o">&amp;</span> <span class="n">e</span><span class="p">){</span>
       <span class="c1">//Partially Handle Runtime Error
</span>       <span class="n">std</span><span class="o">::</span><span class="n">cerr</span> <span class="o">&lt;&lt;</span> <span class="s">"DEBUG ME!!!</span><span class="se">\n</span><span class="s">"</span><span class="p">;</span>
       <span class="k">throw</span><span class="p">;</span> <span class="c1">//Rethrows the same exception.
</span>    <span class="p">}</span></pre></div><div class="step step-level-1" step="12" data-rotate-x="0" data-rotate-y="0" data-rotate-z="0" data-scale="1" data-x="19200" data-y="0" data-z="0"><h1 id="throwing-exceptions">Throwing exceptions</h1><p>Is Easy</p><pre class="highlight code c++"><span class="cp">#include</span> <span class="cpf">&lt;stdexcept&gt;</span><span class="cp">
</span>
<span class="kt">void</span> <span class="nf">foo</span><span class="p">()</span>
<span class="p">{</span>
    <span class="c1">//... do stuff
</span>    <span class="k">if</span><span class="p">(</span><span class="n">bad</span><span class="p">)</span>
        <span class="k">throw</span> <span class="n">std</span><span class="o">::</span><span class="n">runtime_error</span><span class="p">(</span><span class="s">"Weeee!"</span><span class="p">);</span>
<span class="p">}</span></pre></div><div class="step step-level-1" step="13" data-rotate-x="0" data-rotate-y="0" data-rotate-z="0" data-scale="1" data-x="20800" data-y="0" data-z="0"><h1 id="creating-your-own-exceptions">Creating your own exceptions</h1><h2 id="you-can-literally-throw-anything">You can literally throw anything.</h2><pre class="highlight code c++"><span class="k">try</span><span class="p">{</span>
    <span class="k">throw</span> <span class="mi">1</span><span class="p">;</span>
<span class="p">}</span>
<span class="k">catch</span><span class="p">(</span><span class="kt">int</span> <span class="n">e</span><span class="p">){</span>
<span class="c1">// process int
</span><span class="p">}</span></pre><h2 id="but-you-shouldn-t">But you shouldn't.</h2></div><div class="step step-level-1" step="14" data-rotate-x="0" data-rotate-y="0" data-rotate-z="0" data-scale="1" data-x="22400" data-y="0" data-z="0"><h1 id="ok-how-to-really-create-your-own-exceptions">OK How to REALLY create your own exceptions</h1><p>Derive from one of the sub-classes of <tt>std::exception</tt> in <tt>&lt;stdexcept&gt;</tt></p><pre class="highlight code c++"><span class="k">struct</span> <span class="nl">BadFile</span> <span class="p">:</span> <span class="n">std</span><span class="o">::</span><span class="n">runtime_error</span><span class="p">{</span>
    <span class="n">BadFile</span><span class="p">(</span><span class="k">const</span> <span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="o">&amp;</span> <span class="n">what</span><span class="p">)</span><span class="o">:</span><span class="n">std</span><span class="o">::</span><span class="n">runtime_error</span><span class="p">(</span><span class="n">what</span><span class="p">){}</span>
<span class="p">};</span>

<span class="kt">void</span> <span class="nf">foo</span><span class="p">()</span>
<span class="p">{</span>
   <span class="k">throw</span> <span class="n">BadFile</span><span class="p">(</span><span class="s">"This File is Bad"</span><span class="p">);</span>
<span class="p">}</span></pre></div><div class="step step-level-1" step="15" data-rotate-x="0" data-rotate-y="0" data-rotate-z="0" data-scale="1" data-x="24000" data-y="0" data-z="0"><h1 id="the-end">The End</h1></div></div><div id="hovercraft-help"><table><tr><th>Space</th><td>Forward</td></tr><tr><th>Right, Down, Page Down</th><td>Next slide</td></tr><tr><th>Left, Up, Page Up</th><td>Previous slide</td></tr><tr><th>P</th><td>Open presenter console</td></tr><tr><th>H</th><td>Toggle this help</td></tr></table></div><script type="text/javascript" src="js/impress.js"></script><script type="text/javascript" src="js/impressConsole.js"></script><script type="text/javascript" src="js/hovercraft.js"></script></body></html>